lib/linux_admin/ssh_agent.rb

```diff
@@ -0,0 +1,85 @@
+require 'net/ssh'
+require 'fileutils'
+module LinuxAdmin
+  class SshAgent
+    attr_reader :pid
+    attr_reader :sock
+
+    def initialize(agent_socket, ssh_private_key)
+      FileUtils.mkdir_p(File.dirname(agent_socket))
+      agent_details = `ssh-agent -a #{agent_socket}`
+      @sock         = agent_details.split("=")[1].split(" ")[0].chop
+      @pid          = agent_details.split("=")[2].split(" ")[0].chop
+      IO.popen({"SSH_AUTH_SOCK" => @sock, "SSH_AGENT_PID" => @pid}, ["ssh-add", "-"], :mode => 'w') do |f|
+        f.puts(ssh_private_key)
+        if $?.to_i != 0
+          raise StandardError, "Couldn't add key to agent"
+        end
+      end
+    end
+
+    def perform_commands(ip, username, commands, _additional_flags = "")
+      result = nil
+      Net::SSH.start(ip, username, :paranoid => false, :forward_agent => true, :agent_socket_factory => -> { UNIXSocket.open(@sock) }) do |ssh|
```

Comment:
So instead of forcing the user to pass a list we could do something like this which could allow them to do their own error handling rather than not knowing which commands were executed if one fails:

```ruby
def perform_commands(ip, username)
  Net::SSH.start(...) do |ssh|
    yield ssh
  end
end
```

We could also make the commands list optional and check for `block_given?` which would look something like:

```ruby
def perform_commands(ip, username, commands = [])
  Net::SSH.start(...) do |ssh|
    if block_given?
      yield ssh
    else
      commands.each do |cmd|
        ... stuff ...
      end
    end
  end
end
```